### PR TITLE
oom: fix potential race between verification steps

### PR DESCRIPTION
**Pull request by haircommander**
There exists an opportunity for a race between checking the eventfd for events, and checking that the cgroup exists. Specifically:

if we read the eventfd before the cgroup is destroyed, but check that the cgroup exists after its destroyed, we will miss a real oom (because the first event triggered was an oom)

To mitigate this, we need to check the existance of the file before we read the event. Since the kernel first removes the files, then triggers an event, we resolve the race here

Signed-off-by: Peter Hunt <pehunt@redhat.com>

### CODE REVIEW

**Comment by haircommander** [comment]
@giuseppe PTAL

File: src/conmon.c [hunk 1]
```diff
@@ -570,21 +570,30 @@ static gboolean oom_cb_cgroup_v1(int fd, GIOCondition condition, gpointer user_d
...
-		if (event_count == 1)
-			return G_SOURCE_CONTINUE;
+		if (event_count == 1) {
+			/* Before we decide the only event was the cgroup deletion, let's double-check for another
+			 * event in the event_fd. If we find one, the cgroup could have been deleted after we read the
+			 * first time, but before we checked for the event_control file.
+			 */
+			if (read(fd, &event_count, sizeof(uint64_t)) < 0)
+				return G_SOURCE_CONTINUE;
+
```
Comment (giuseppe) [inline src/conmon.c]: it looks correct to me, but should we ensure the fd is set to not blocking?

Comment (haircommander) [reply src/conmon.c]: whoops, I thought it was but it wasn't. good catch

**Comment by haircommander** [comment]
PR simplified. We can check for the cgroup existing first, then read from the fd. Since a removal happens before the event is triggered, we still avoid the race, and this version cleans the code up a bit.

PTAL @giuseppe @rhatdan @mrunalp @saschagrunert @mheon

File: src/conmon.c (hunk 1 above)
Comment (haircommander) [reply src/conmon.c]: With the new version, we don't need a non-block fd anymore

**Comment by mheon** [comment]
LGTM

**Comment by rhatdan** [comment]
@giuseppe Please review and merge.

**Pull request merged**